dg.go

```diff
@@ -1,315 +1,480 @@
-    d.dg.lock.Lock()
-    defer d.dg.lock.Unlock()
-    if d.deleted {
-        return &ErrNodeDeleted{
-            d.id,
-        }
-    }
-    if _, ok := d.dg.nodes[toNodeID]; !ok {
-        return &ErrNodeNotFound{
-            toNodeID,
-        }
-    }
-    if _, ok := d.dg.connectionsFromNode[d.id][toNodeID]; !ok {
-        return &ErrConnectionDoesNotExist{
-            d.id,
-            toNodeID,
-        }
-    }
-    delete(d.dg.connectionsFromNode[d.id], toNodeID)
-    delete(d.dg.connectionsToNode[toNodeID], d.id)
-    return nil
+func (n *node[NodeType]) ListOutboundConnections() (map[string]Node[NodeType], error) {
+	n.dg.lock.Lock()
+	defer n.dg.lock.Unlock()
+	if n.deleted {
+		return nil, &ErrNodeDeleted{n.id}
+	}
+	result := make(map[string]Node[NodeType], len(n.dg.connectionsFromNode[n.id]))
+	for toNodeID := range n.dg.connectionsFromNode[n.id] {
+		result[toNodeID] = n.dg.nodes[toNodeID]
+	}
+	return result, nil
 }
 
-func (d *node[NodeType]) Remove() error {
-    d.dg.lock.Lock()
-    defer d.dg.lock.Unlock()
-    if d.deleted {
-        return &ErrNodeDeleted{
-            d.id,
-        }
-    }
-    for toNodeID := range d.dg.connectionsFromNode[d.id] {
-        delete(d.dg.connectionsToNode[toNodeID], d.id)
-    }
-    delete(d.dg.connectionsFromNode, d.id)
-    for fromNodeID := range d.dg.connectionsToNode[d.id] {
-        delete(d.dg.connectionsFromNode[fromNodeID], d.id)
-    }
-    delete(d.dg.connectionsToNode, d.id)
-    delete(d.dg.nodes, d.id)
-    d.deleted = true
-    return nil
+// dependencyResolved is used to notify a node that one of its dependencies have had their resolution
+// status set. Once all dependencies are resolved, the node is set as finalized (ready for processing).
+// Caller should have appropriate mutex locked before calling.
+func (n *node[NodeType]) dependencyResolved(dependencyNodeID string, dependencyResolution ResolutionStatus) error {
+	if n.deleted {
+		return &ErrNodeDeleted{n.id}
+	}
+	if dependencyResolution == Waiting {
+		// Illegal state
+		return ErrNotifiedOfWaiting{n.id, dependencyNodeID}
+	}
+	dependencyType, isOutstandingDependency := n.outstandingDependencies[dependencyNodeID]
+	if !isOutstandingDependency {
+		// Now determine if the missing item was because the dependency was already resolved, or
+		// because there was never a connection.
+		_, isConnected := n.dg.connectionsToNode[n.id][dependencyNodeID]
+		if isConnected {
+			// As designed, this is an internal function. So we guard against this in resolveNode.
+			panic(ErrDuplicateDependencyResolution{n.id, dependencyNodeID})
+		} else {
+			panic(ErrConnectionDoesNotExist{dependencyNodeID, n.id})
+		}
+	}
+	delete(n.outstandingDependencies, dependencyNodeID)
+	if dependencyType == ObviatedDependency {
+		return nil // Nothing to do.
+	}
+	// If resolution is unresolvable, fail if current type is AND, or if there are no remaining OR dependencies.
```

Comment:
"If the resolution is unresolvable" there's no resolution. Maybe if the "dependency" is unresolvable? Or just "unresolved" as it might be "resolvable" in the future?

Reply:
I updated some of the relevant comments.